In [ ]:
# Drop rows where BPL is greater than 56
df_filtered = df_filtered[df_filtered['BPL'] <= 56].copy()

# Display the updated dataframe
display(df_filtered.head())

,YEAR,SAMPLE,SERIAL,CBSERIAL,HHWT,STATEFIP,PERNUM,PERWT,SEX,AGE,RACE,HISPAN,BPL,EDUC,EDUCD,birth year
13,2006,200601,6,464,62.0,1,1,63.0,2,23,1,0,47,10,101,1983
17,2006,200601,8,555,98.0,1,3,154.0,1,22,1,0,1,7,71,1984
30,2006,200601,16,1139,46.0,1,2,109.0,2,25,1,0,1,10,101,1981
113,2006,200601,59,4188,232.0,1,3,353.0,1,23,1,0,48,5,50,1983
144,2006,200601,76,5305,60.0,1,3,138.0,1,24,1,0,37,7,71,1982


In [ ]:
import pandas as pd

file_path = '/content/drive/MyDrive/Colab Notebooks/Thesis/Data/Mechanisms Data /usa_00003.csv'
df = pd.read_csv(file_path)

# Display the first few rows of the dataframe
display(df.head())

,YEAR,SAMPLE,SERIAL,CBSERIAL,HHWT,STATEFIP,PERNUM,PERWT,SEX,AGE,RACE,RACED,HISPAN,HISPAND,BPL,BPLD,EDUC,EDUCD
0,2006,200601,1,91,88.0,1,1,87.0,1,47,1,100,0,0,1,100,6,65
1,2006,200601,2,280,105.0,1,1,106.0,2,59,1,100,0,0,47,4700,6,62
2,2006,200601,2,280,105.0,1,2,107.0,1,57,1,100,0,0,1,100,6,62
3,2006,200601,2,280,105.0,1,3,114.0,1,30,1,100,0,0,48,4800,7,71
4,2006,200601,2,280,105.0,1,4,150.0,1,50,1,100,0,0,47,4700,4,40


In [ ]:
# Create 'birth year' column
df['birth year'] = df['YEAR'] - df['AGE']

# Filter for birth years 1981, 1982, 1983, 1984
birth_years_to_keep = [1981, 1982, 1983, 1984]
df_filtered = df[df['birth year'].isin(birth_years_to_keep)].copy()

# Drop columns RACED, HISPAND, BPLD
columns_to_drop = ['RACED', 'HISPAND', 'BPLD']
df_filtered = df_filtered.drop(columns=columns_to_drop, errors='ignore')

# Display the new dataframe
display(df_filtered.head())

,YEAR,SAMPLE,SERIAL,CBSERIAL,HHWT,STATEFIP,PERNUM,PERWT,SEX,AGE,RACE,HISPAN,BPL,EDUC,EDUCD,birth year
13,2006,200601,6,464,62.0,1,1,63.0,2,23,1,0,47,10,101,1983
17,2006,200601,8,555,98.0,1,3,154.0,1,22,1,0,1,7,71,1984
30,2006,200601,16,1139,46.0,1,2,109.0,2,25,1,0,1,10,101,1981
113,2006,200601,59,4188,232.0,1,3,353.0,1,23,1,0,48,5,50,1983
144,2006,200601,76,5305,60.0,1,3,138.0,1,24,1,0,37,7,71,1982


In [ ]:
import numpy as np

# Drop STATEFIP and rename BPL
df_filtered = df_filtered.drop(columns=['STATEFIP'], errors='ignore')
df_filtered = df_filtered.rename(columns={'BPL': 'STATE_BPL'})

# Drop invalid EDUCD (0, 1, 999)
df_filtered = df_filtered[~df_filtered['EDUCD'].isin([0, 1, 999])].copy()

# Binary Outcomes
# HS Grad: EDUCD between 62 and 116
df_filtered['hs_grad'] = ((df_filtered['EDUCD'] >= 62) & (df_filtered['EDUCD'] <= 116)).astype(int)

# College Grad: EDUCD between 101 and 116
df_filtered['college_grad'] = ((df_filtered['EDUCD'] >= 101) & (df_filtered['EDUCD'] <= 116)).astype(int)

# Some College: EDUCD between 65 and 116
df_filtered['some_college'] = ((df_filtered['EDUCD'] >= 65) & (df_filtered['EDUCD'] <= 116)).astype(int)

# Years of Education Mapping
def map_educd_to_years(educd):
    if educd <= 2: return 0
    elif educd == 10: return 2.5
    elif educd == 14: return 1
    elif educd == 15: return 2
    elif educd == 16: return 3
    elif educd == 17: return 4
    elif educd == 21: return 5.5
    elif educd == 22: return 5
    elif educd == 23: return 6
    elif educd == 24: return 7.5
    elif educd == 25: return 7
    elif educd == 26: return 8
    elif educd == 30: return 9
    elif educd == 40: return 10
    elif educd == 50: return 11
    elif educd in [60, 61, 62, 63, 64]: return 12
    elif educd == 65: return 12.5
    elif educd in [70, 71]: return 13
    elif educd in [80, 81, 82, 83]: return 14
    elif educd == 90: return 15
    elif educd in [100, 101]: return 16
    elif educd == 110: return 17
    elif educd == 111: return 18
    elif educd == 112: return 19
    elif educd == 113: return 20
    elif educd == 114: return 18
    elif educd == 115: return 19
    elif educd == 116: return 20
    else: return np.nan

df_filtered['years_of_education'] = df_filtered['EDUCD'].apply(map_educd_to_years)

# Display the modified dataframe
display(df_filtered.head())

,YEAR,SAMPLE,SERIAL,CBSERIAL,HHWT,PERNUM,PERWT,SEX,AGE,RACE,HISPAN,STATE_BPL,EDUC,EDUCD,birth year,hs_grad,college_grad,some_college,years_of_education
13,2006,200601,6,464,62.0,1,63.0,2,23,1,0,47,10,101,1983,1,1,1,16.0
17,2006,200601,8,555,98.0,3,154.0,1,22,1,0,1,7,71,1984,1,0,1,13.0
30,2006,200601,16,1139,46.0,2,109.0,2,25,1,0,1,10,101,1981,1,1,1,16.0
113,2006,200601,59,4188,232.0,3,353.0,1,23,1,0,48,5,50,1983,0,0,0,11.0
144,2006,200601,76,5305,60.0,3,138.0,1,24,1,0,37,7,71,1982,1,0,1,13.0


# Task
Load the Medicaid eligibility dataset and merge it with the processed education dataframe `df_filtered` based on 'STATE_BPL' and 'birth year'. Please check the directory `/content/drive/MyDrive/Colab Notebooks/Thesis/Data/` or `/content/drive/MyDrive/Colab Notebooks/Thesis/Data/Mechanisms Data /` for the appropriate Medicaid file. If the file is not found, list the contents of these directories to identify the correct file.

# Task
Load the file `[Final]_US_Crime_Rates_1999_2010.csv` from the directory `/content/drive/MyDrive/Colab Notebooks/Thesis/Data/` into a pandas DataFrame. Display the first few rows and the column names of this new dataframe to identify the relevant columns for merging (likely state and year information). Then, rename the columns in this new dataframe to match `STATE_BPL` and `birth year` in `df_filtered` if necessary, ensuring data types are consistent. Finally, merge `df_filtered` with this loaded dataframe and display the head of the merged result.

## Load Medicaid/Crime Data

### Subtask:
Load the US Crime Rates CSV file into a pandas DataFrame and inspect its initial rows and columns.


**Reasoning**:
Load the US Crime Rates dataset from the specified path and inspect its structure by displaying the first few rows and column names.



In [ ]:
file_path_crime = '/content/drive/MyDrive/Colab Notebooks/Thesis/Data/[Final]_US_Crime_Rates_1999_2010.csv'
df_crime = pd.read_csv(file_path_crime)

# Display the first few rows of the dataframe
display(df_crime.head())

# Print the column names
print(df_crime.columns)

,fips,year,age,crime_count,brthyr,Age,State,Population,Drug_Rate,Financial_Rate,...,Violent_Rate,cumulative_eligibility,elig_0_5,elig_6_11,elig_12_18,trend_cohort,Childhood_Avg_Unemployment,economically_motivated_crimes,state_abb,childhood_avg_income
0,1,2000,19,1594,1981,19,AL,67769,142.395491,57.400877,...,168.661187,3.228658,1.000340,0.858396,1.369921,0,7.957895,235.210790,AL,10537
1,1,2001,19,1741,1982,19,AL,65031,141.009672,66.122311,...,182.374560,3.559218,0.963729,0.939953,1.655536,1,7.642105,267.718473,AL,10732
2,1,2001,20,1686,1981,20,AL,68511,133.263272,87.285253,...,178.949366,3.228658,1.000340,0.858396,1.369921,0,7.957895,246.091868,AL,10537
3,1,2002,19,1758,1983,19,AL,64169,154.124266,55.946018,...,213.342891,3.929335,0.927237,0.999889,2.002209,2,7.168421,273.964064,AL,10937
4,1,2002,20,1655,1982,20,AL,65515,152.636801,69.449744,...,216.896894,3.559218,0.963729,0.939953,1.655536,1,7.642105,252.613905,AL,10732


Index(['fips', 'year', 'age', 'crime_count', 'brthyr', 'Age', 'State',
       'Population', 'Drug_Rate', 'Financial_Rate', 'Property_Rate',
       'PublicOrder_Rate', 'Violent_Rate', 'cumulative_eligibility',
       'elig_0_5', 'elig_6_11', 'elig_12_18', 'trend_cohort',
       'Childhood_Avg_Unemployment', 'economically_motivated_crimes',
       'state_abb', 'childhood_avg_income'],
      dtype='object')


**Reasoning**:
Rename the matching columns in the crime dataframe and merge it with the filtered dataframe based on 'STATE_BPL' and 'birth year', ensuring data consistency.



In [ ]:
df_crime = df_crime.rename(columns={'fips': 'STATE_BPL', 'brthyr': 'birth year'})

# Ensure merge keys are integers
df_filtered['STATE_BPL'] = df_filtered['STATE_BPL'].astype(int)
df_filtered['birth year'] = df_filtered['birth year'].astype(int)
df_crime['STATE_BPL'] = df_crime['STATE_BPL'].astype(int)
df_crime['birth year'] = df_crime['birth year'].astype(int)

# Merge dataframes
df_merged = pd.merge(df_filtered, df_crime, on=['STATE_BPL', 'birth year'], how='left')

# Display the merged dataframe
display(df_merged.head())

,YEAR,SAMPLE,SERIAL,CBSERIAL,HHWT,PERNUM,PERWT,SEX,AGE,RACE,...,Violent_Rate,cumulative_eligibility,elig_0_5,elig_6_11,elig_12_18,trend_cohort,Childhood_Avg_Unemployment,economically_motivated_crimes,state_abb,childhood_avg_income
0,2006,200601,6,464,62.0,1,63.0,2,23,1,...,171.149766,4.143075,0.71315,0.705927,2.723999,2.0,6.052632,271.400074,TN,11747.0
1,2006,200601,6,464,62.0,1,63.0,2,23,1,...,193.022283,4.143075,0.71315,0.705927,2.723999,2.0,6.052632,287.504280,TN,11747.0
2,2006,200601,6,464,62.0,1,63.0,2,23,1,...,198.456725,4.143075,0.71315,0.705927,2.723999,2.0,6.052632,270.135504,TN,11747.0
3,2006,200601,6,464,62.0,1,63.0,2,23,1,...,208.164069,4.143075,0.71315,0.705927,2.723999,2.0,6.052632,259.497255,TN,11747.0
4,2006,200601,6,464,62.0,1,63.0,2,23,1,...,204.545455,4.143075,0.71315,0.705927,2.723999,2.0,6.052632,258.797654,TN,11747.0


## Summary:

Here is the summary of the data analysis task:

### Data Analysis Key Findings
*   **Dataset Integration:** A US Crime Rates dataset spanning 1999-2010 was successfully loaded and prepared for integration.
*   **Column Mapping:** The dataset contained key identifier columns `fips` and `brthyr`, which were renamed to `STATE_BPL` and `birth year` respectively to match the primary dataset.
*   **Data Consistency:** Merge keys (`STATE_BPL` and `birth year`) were standardized as integers across both dataframes to ensure accurate linking.
*   **Enriched Dataset:** The final merged dataframe combines the original demographic data with crime statistics (such as `Drug_Rate`, `Violent_Rate`) and economic indicators, enabling a multifaceted analysis of how these environmental factors correlate with the existing birth and location data.

### Insights or Next Steps
*   With the crime and economic data now attached to specific state-year cohorts, the analysis can proceed to investigate correlations between early-life environmental factors (like crime rates) and later-life outcomes.
*   It is advisable to check for missing values in the merged columns, as the left join may have resulted in null values for years or states not present in the crime dataset.


In [ ]:
# Calculate the range of the AGE column
min_age = df_merged['AGE'].min()
max_age = df_merged['AGE'].max()

print(f"The ages range from {min_age} to {max_age}.")

# Display unique ages to see the distribution
print("Unique ages present:", sorted(df_merged['AGE'].unique()))

The ages range from 22 to 39.
Unique ages present: [np.int64(22), np.int64(23), np.int64(24), np.int64(25), np.int64(26), np.int64(27), np.int64(28), np.int64(29), np.int64(30), np.int64(31), np.int64(32), np.int64(33), np.int64(34), np.int64(35), np.int64(36), np.int64(37), np.int64(38), np.int64(39)]


In [ ]:
# Filter for ages 28-34 (inclusive)
df_merged = df_merged[(df_merged['AGE'] >= 28) & (df_merged['AGE'] <= 34)].copy()

# Display the number of rows after filtering
print(f"Rows remaining after filtering for ages 28-34: {len(df_merged)}")

# Check the age range again to verify
print(f"New age range: {df_merged['AGE'].min()} - {df_merged['AGE'].max()}")

Rows remaining after filtering for ages 28-34: 4927309
New age range: 28 - 34


In [ ]:
# Calculate number of rows in the unmerged dataframe
num_rows_unmerged = len(df_filtered)
print(f"The unmerged dataframe (df_filtered) has {num_rows_unmerged} rows.")

The unmerged dataframe (df_filtered) has 1783102 rows.


In [ ]:
# Select only the relevant columns for Medicaid eligibility + merge keys
medicaid_cols = ['STATE_BPL', 'birth year', 'cumulative_eligibility', 'elig_0_5', 'elig_6_11', 'elig_12_18']

# Create a subset of the crime dataframe with only these columns
df_medicaid = df_crime[medicaid_cols].copy()

# Drop duplicates to ensure one row per State-Birth Year
# We take the first occurrence since the eligibility values should be constant for the cohort
df_medicaid_unique = df_medicaid.drop_duplicates(subset=['STATE_BPL', 'birth year'])

# Merge this unique Medicaid data back into the original filtered dataframe
df_merged_final = pd.merge(df_filtered, df_medicaid_unique, on=['STATE_BPL', 'birth year'], how='left')

# Display the head of the new merged dataframe
display(df_merged_final.head())

# Verify row counts
print(f"Rows in original df_filtered: {len(df_filtered)}")
print(f"Rows in new merged dataframe: {len(df_merged_final)}")

,YEAR,SAMPLE,SERIAL,CBSERIAL,HHWT,PERNUM,PERWT,SEX,AGE,RACE,...,EDUCD,birth year,hs_grad,college_grad,some_college,years_of_education,cumulative_eligibility,elig_0_5,elig_6_11,elig_12_18
0,2006,200601,6,464,62.0,1,63.0,2,23,1,...,101,1983,1,1,1,16.0,4.143075,0.713150,0.705927,2.723999
1,2006,200601,8,555,98.0,3,154.0,1,22,1,...,71,1984,1,0,1,13.0,4.904218,0.901487,1.445018,2.557714
2,2006,200601,16,1139,46.0,2,109.0,2,25,1,...,101,1981,1,1,1,16.0,3.228658,1.000340,0.858396,1.369921
3,2006,200601,59,4188,232.0,3,353.0,1,23,1,...,50,1983,0,0,0,11.0,3.879248,1.260066,0.970312,1.648869
4,2006,200601,76,5305,60.0,3,138.0,1,24,1,...,71,1982,1,0,1,13.0,3.370484,0.994488,0.791701,1.584296


Rows in original df_filtered: 1783102
Rows in new merged dataframe: 1783102


In [ ]:
output_path = '/content/drive/MyDrive/Colab Notebooks/Thesis/Data/FINAL - Cleaned Education data.csv'
df_merged_final.to_csv(output_path, index=False)
print(f"Dataframe saved to: {output_path}")

Dataframe saved to: /content/drive/MyDrive/Colab Notebooks/Thesis/Data/FINAL - Cleaned Education data.csv


In [ ]:
# Filter for ages 28-34 (inclusive)
df_merged_final = df_merged_final[(df_merged_final['AGE'] >= 28) & (df_merged_final['AGE'] <= 34)].copy()

# Display the number of rows after filtering
print(f"Rows remaining after filtering for ages 28-34: {len(df_merged_final)}")

# Check the age range again to verify
print(f"New age range: {df_merged_final['AGE'].min()} - {df_merged_final['AGE'].max()}")

Rows remaining after filtering for ages 28-34: 849173
New age range: 28 - 34


# Task
Perform a regression analysis to evaluate the impact of Medicaid eligibility on educational outcomes.

Using the `df_merged_final` dataframe:
1.  **Import Libraries**: Import `statsmodels.formula.api` as `smf` and `statsmodels.api` as `sm`.
2.  **Define Regression Function**: Create a function that accepts an `outcome` variable name and an `eligibility` variable name. Inside the function:
    *   Construct a regression formula: `f"{outcome} ~ {eligibility} + C(STATE_BPL) + C(birth_year) + C(AGE) + C(SEX) + C(RACE) + C(HISPAN)"`. Note: Use `birth year` (with space) if column name has space, or rename/handle appropriately (e.g., `Q("birth year")`).
    *   Fit a Weighted Least Squares (WLS) model using `PERWT` as weights.
    *   Calculate robust standard errors clustered by `STATE_BPL`.
    *   Return the coefficient, standard error, and p-value for the eligibility variable.
3.  **Run Regressions**: Iterate through the 4 outcomes (`hs_grad`, `college_grad`, `years_of_education`, `some_college`) and the 4 eligibility measures (`elig_0_5`, `elig_6_11`, `elig_12_18`, `cumulative_eligibility`).
4.  **Display Results**: Compile the results into a pandas DataFrame and display it to show the impact of each eligibility window on each educational outcome.
5.  **Summarize**: Briefly summarize the findings in natural language.

## Import Libraries

### Subtask:
Import the necessary statsmodels libraries for regression analysis.


**Reasoning**:
Import the required statsmodels libraries for regression analysis as per the instructions.



In [ ]:
import statsmodels.formula.api as smf
import statsmodels.api as sm

print("Libraries imported.")

Libraries imported.


## Define Regression Function

### Subtask:
Create a function to perform WLS regression with clustered standard errors.


**Reasoning**:
I will first rename the 'birth year' column to 'birth_year' to avoid syntax issues in the statsmodels formula. Then, I will define the `run_regression` function as requested, which performs WLS regression with clustered standard errors.



In [ ]:
# Rename 'birth year' to 'birth_year' for easier use in formulas
df_merged_final = df_merged_final.rename(columns={'birth year': 'birth_year'})

def run_regression(outcome_var, eligibility_var, data):
    """
    Runs a WLS regression of outcome_var on eligibility_var with controls and fixed effects.

    Args:
    outcome_var (str): Name of the outcome variable.
    eligibility_var (str): Name of the eligibility variable.
    data (pd.DataFrame): The dataframe containing the data.

    Returns:
    tuple: (coefficient, standard error, p-value) for the eligibility variable.
    """

    # Construct the formula
    # Note: We use C() to treat variables as categorical fixed effects
    formula = f"{outcome_var} ~ {eligibility_var} + C(STATE_BPL) + C(birth_year) + C(AGE) + C(SEX) + C(RACE) + C(HISPAN)"

    # Fit the WLS model
    # We use 'PERWT' as weights (Person Weights)
    model = smf.wls(formula=formula, data=data, weights=data['PERWT'])

    # Fit with clustered standard errors on 'STATE_BPL'
    results = model.fit(cov_type='cluster', cov_kwds={'groups': data['STATE_BPL']})

    # Extract results for the eligibility variable
    coef = results.params[eligibility_var]
    std_err = results.bse[eligibility_var]
    p_value = results.pvalues[eligibility_var]

    return coef, std_err, p_value

print("Function 'run_regression' defined.")
print("Column 'birth year' renamed to 'birth_year'.")

Function 'run_regression' defined.
Column 'birth year' renamed to 'birth_year'.


**Reasoning**:
Iterate through the specified outcome and eligibility variables, running the regression for each combination. Store the results (coefficient, standard error, p-value) in a list and then convert it to a DataFrame for display.



# Task
Check for missing values in the `df_merged_final` dataframe, specifically in the outcome variables (`hs_grad`, `college_grad`, `years_of_education`, `some_college`), eligibility measures (`elig_0_5`, `elig_6_11`, `elig_12_18`, `cumulative_eligibility`), and control variables (`STATE_BPL`, `birth_year`, `AGE`, `SEX`, `RACE`, `HISPAN`, `PERWT`). Display the count of missing values for each column. Then, drop rows with any missing values in these specific columns to create a clean dataframe `df_clean`. Finally, re-run the regression loop using `df_clean` and display the results dataframe. Summarize the findings.

## Check for Missing Values

### Subtask:
Identify and count missing values in the relevant columns of the dataframe.


**Reasoning**:
I will define the list of relevant columns including outcomes, eligibility measures, and controls, and then calculate and display the count of missing values for each to identify data quality issues.



In [ ]:
# Define relevant columns
relevant_columns = [
    'hs_grad', 'college_grad', 'years_of_education', 'some_college',
    'elig_0_5', 'elig_6_11', 'elig_12_18', 'cumulative_eligibility',
    'STATE_BPL', 'birth_year', 'AGE', 'SEX', 'RACE', 'HISPAN', 'PERWT'
]

# Check for missing values in these columns
missing_values = df_merged_final[relevant_columns].isnull().sum()

# Display the count of missing values
print("Missing values per column:")
print(missing_values[missing_values > 0])

# Also print total rows for context
print(f"\nTotal rows in dataframe: {len(df_merged_final)}")

Missing values per column:
years_of_education          253
elig_0_5                  32438
elig_6_11                 32438
elig_12_18                32438
cumulative_eligibility    32438
dtype: int64

Total rows in dataframe: 849173


**Reasoning**:
I will drop rows with missing values in the relevant columns to create a clean dataset, then iterate through combinations of outcome and eligibility variables to run WLS regressions, collecting the results into a DataFrame.



In [ ]:
# Drop rows with missing values in the relevant columns
df_clean = df_merged_final.dropna(subset=relevant_columns).copy()

# Print the number of rows in the clean dataframe
print(f"Rows in clean dataframe: {len(df_clean)}")

# Initialize a list to store results
results = []

# Define outcomes and eligibility measures
outcomes = ['hs_grad', 'college_grad', 'years_of_education', 'some_college']
eligibility_measures = ['elig_0_5', 'elig_6_11', 'elig_12_18', 'cumulative_eligibility']

# Iterate through each outcome and eligibility measure
for outcome in outcomes:
    for eligibility in eligibility_measures:
        # Run the regression
        coef, std_err, p_value = run_regression(outcome, eligibility, df_clean)

        # Append the results
        results.append({
            'Outcome': outcome,
            'Eligibility Variable': eligibility,
            'Coefficient': coef,
            'Standard Error': std_err,
            'P-value': p_value
        })

# Convert results to a DataFrame
results_df = pd.DataFrame(results)

# Display the results
display(results_df)

Rows in clean dataframe: 816497


KeyboardInterrupt: 

# Task
Run the regression analysis again using the cleaned dataframe `df_clean`. Iterate through the 4 outcomes (`hs_grad`, `college_grad`, `years_of_education`, `some_college`) and the 4 eligibility measures (`elig_0_5`, `elig_6_11`, `elig_12_18`, `cumulative_eligibility`). For each combination, run the WLS regression using the `run_regression` function defined previously. Store the results (Outcome, Eligibility Variable, Coefficient, Standard Error, P-value) in a DataFrame and display it. Finally, provide a brief summary of the findings based on the regression results.

## Display Clean Data Header

### Subtask:
Inspect the first few rows of the cleaned dataframe.


**Reasoning**:
The user wants to inspect the cleaned dataframe `df_clean`. I will display the first 5 rows using the `.head()` method.



In [ ]:
display(df_clean.head())

,YEAR,SAMPLE,SERIAL,CBSERIAL,HHWT,PERNUM,PERWT,SEX,AGE,RACE,...,EDUCD,birth_year,hs_grad,college_grad,some_college,years_of_education,cumulative_eligibility,elig_0_5,elig_6_11,elig_12_18
341272,2009,200901,42,2992,126.0,3,94.0,1,28,1,...,63,1981,1,0,0,12.0,1.828912,0.650532,0.612582,0.565799
341273,2009,200901,48,3322,110.0,1,111.0,1,28,1,...,63,1981,1,0,0,12.0,3.252349,0.905107,0.649300,1.697942
341274,2009,200901,48,3322,110.0,2,110.0,2,28,1,...,63,1981,1,0,0,12.0,3.142020,0.998988,0.945851,1.197181
341275,2009,200901,68,4312,263.0,1,263.0,2,28,2,...,63,1981,1,0,0,12.0,3.228658,1.000340,0.858396,1.369921
341281,2009,200901,107,6937,27.0,2,33.0,1,28,1,...,81,1981,1,0,1,14.0,3.228658,1.000340,0.858396,1.369921


**Reasoning**:
I will re-run the regression analysis on the cleaned dataframe `df_clean`. I'll iterate through the defined outcomes and eligibility measures, applying the `run_regression` function to each combination. The results (coefficient, standard error, and p-value) will be collected and compiled into a summary DataFrame for display.



In [ ]:
# Initialize a list to store results
results = []

# Define outcomes and eligibility measures
outcomes = ['hs_grad', 'college_grad', 'years_of_education', 'some_college']
eligibility_measures = ['elig_0_5', 'elig_6_11', 'elig_12_18', 'cumulative_eligibility']

print("Starting regression analysis...")

# Iterate through each outcome and eligibility measure
for outcome in outcomes:
    for eligibility in eligibility_measures:
        print(f"Running regression for Outcome: {outcome}, Eligibility: {eligibility}...")
        # Run the regression
        coef, std_err, p_value = run_regression(outcome, eligibility, df_clean)

        # Append the results
        results.append({
            'Outcome': outcome,
            'Eligibility Variable': eligibility,
            'Coefficient': coef,
            'Standard Error': std_err,
            'P-value': p_value
        })

# Convert results to a DataFrame
results_df = pd.DataFrame(results)

# Display the results
print("\nRegression Analysis Results:")
display(results_df)

Starting regression analysis...
Running regression for Outcome: hs_grad, Eligibility: elig_0_5...
Running regression for Outcome: hs_grad, Eligibility: elig_6_11...
Running regression for Outcome: hs_grad, Eligibility: elig_12_18...
Running regression for Outcome: hs_grad, Eligibility: cumulative_eligibility...
Running regression for Outcome: college_grad, Eligibility: elig_0_5...
Running regression for Outcome: college_grad, Eligibility: elig_6_11...
Running regression for Outcome: college_grad, Eligibility: elig_12_18...
Running regression for Outcome: college_grad, Eligibility: cumulative_eligibility...
Running regression for Outcome: years_of_education, Eligibility: elig_0_5...
Running regression for Outcome: years_of_education, Eligibility: elig_6_11...
Running regression for Outcome: years_of_education, Eligibility: elig_12_18...
Running regression for Outcome: years_of_education, Eligibility: cumulative_eligibility...
Running regression for Outcome: some_college, Eligibility: el

,Outcome,Eligibility Variable,Coefficient,Standard Error,P-value
0,hs_grad,elig_0_5,0.002412,0.010933,0.825408
1,hs_grad,elig_6_11,-0.001189,0.003814,0.755289
2,hs_grad,elig_12_18,-0.002215,0.001866,0.235182
3,hs_grad,cumulative_eligibility,-0.002324,0.002183,0.287198
4,college_grad,elig_0_5,-0.004627,0.011685,0.692131
5,college_grad,elig_6_11,0.006103,0.007758,0.431434
6,college_grad,elig_12_18,0.002607,0.002376,0.272572
7,college_grad,cumulative_eligibility,0.003809,0.002558,0.136377
8,years_of_education,elig_0_5,0.028997,0.067768,0.668735
9,years_of_education,elig_6_11,0.015274,0.041952,0.715795


# Task
Prepare the economic control variables for the analysis. First, select the columns `Childhood_Avg_Unemployment` and `childhood_avg_income` along with the merge keys `STATE_BPL` and `birth year` from the dataframe `df_crime_aggregated`. Create a new column `log_childhood_income` by applying a natural log transformation to `childhood_avg_income`. Then, rename the `birth year` column to `birth_year` to match the schema of `df_clean`. Merge these new economic variables into `df_clean` using a left join on `STATE_BPL` and `birth_year`. Finally, check for any missing values in the newly added columns and display the first few rows of the updated dataframe to verify the merge. Summarize the successful addition of these economic controls.

## Prepare Economic Data

### Subtask:
Select relevant economic columns from the aggregated crime dataframe.


**Reasoning**:
I will select the specified columns from `df_crime_aggregated` to create `df_economic` and display its first few rows.



In [ ]:
# Select relevant economic columns
economic_cols = ['STATE_BPL', 'birth year', 'Childhood_Avg_Unemployment', 'childhood_avg_income']
df_economic = df_crime_aggregated[economic_cols].copy()

# Display the new dataframe
display(df_economic.head())

,STATE_BPL,birth year,Childhood_Avg_Unemployment,childhood_avg_income
0,1,1981,7.957895,10537.0
1,1,1982,7.642105,10732.0
2,1,1983,7.168421,10937.0
3,1,1984,6.752632,11137.0
4,2,1981,8.442105,15957.0


**Reasoning**:
I will create the `log_childhood_income` column, rename `birth year` to `birth_year` for consistency, and merge the economic data into `df_clean`. Finally, I will check for missing values in the new columns and display the updated dataframe.



In [ ]:
import numpy as np

# Create log income variable
df_economic['log_childhood_income'] = np.log(df_economic['childhood_avg_income'])

# Rename 'birth year' to 'birth_year' to match df_clean
df_economic = df_economic.rename(columns={'birth year': 'birth_year'})

# Merge economic variables into df_clean
# We use left join to keep all rows in the clean analysis set
df_clean = pd.merge(df_clean, df_economic, on=['STATE_BPL', 'birth_year'], how='left')

# Check for missing values in the new columns
new_cols = ['Childhood_Avg_Unemployment', 'childhood_avg_income', 'log_childhood_income']
print("Missing values in added economic columns:")
print(df_clean[new_cols].isnull().sum())

# Display the updated dataframe
display(df_clean.head())

Missing values in added economic columns:
Childhood_Avg_Unemployment    0
childhood_avg_income          0
log_childhood_income          0
dtype: int64


,YEAR,SAMPLE,SERIAL,CBSERIAL,HHWT,PERNUM,PERWT,SEX,AGE,RACE,...,college_grad,some_college,years_of_education,cumulative_eligibility,elig_0_5,elig_6_11,elig_12_18,Childhood_Avg_Unemployment,childhood_avg_income,log_childhood_income
0,2009,200901,42,2992,126.0,3,94.0,1,28,1,...,0,0,12.0,1.828912,0.650532,0.612582,0.565799,8.189474,9177.0,9.124456
1,2009,200901,48,3322,110.0,1,111.0,1,28,1,...,0,0,12.0,3.252349,0.905107,0.649300,1.697942,5.705263,11898.0,9.384126
2,2009,200901,48,3322,110.0,2,110.0,2,28,1,...,0,0,12.0,3.142020,0.998988,0.945851,1.197181,6.610526,12032.0,9.395325
3,2009,200901,68,4312,263.0,1,263.0,2,28,2,...,0,0,12.0,3.228658,1.000340,0.858396,1.369921,7.957895,10537.0,9.262648
4,2009,200901,107,6937,27.0,2,33.0,1,28,1,...,0,1,14.0,3.228658,1.000340,0.858396,1.369921,7.957895,10537.0,9.262648


# Task
Create a function `run_economic_trend_regression` that fits the model: `outcome ~ eligibility + Childhood_Avg_Unemployment + log_childhood_income + YEAR + C(STATE_BPL) + C(AGE) + C(SEX) + C(RACE) + C(HISPAN)`. Use WLS with `PERWT` and cluster errors by `STATE_BPL`. Then, iterate through the 4 outcomes (`hs_grad`, `college_grad`, `years_of_education`, `some_college`) and 4 eligibility measures (`elig_0_5`, `elig_6_11`, `elig_12_18`, `cumulative_eligibility`) using the `df_clean` dataframe. Store the results (Outcome, Eligibility, Coefficients for Eligibility, Unemployment, Log Income, Year, SE, P-value) in a DataFrame and display it. Finally, summarize the findings from this new specification.

## Define Updated Regression Function

### Subtask:
Create a function to run WLS regression with additional economic and time controls.


**Reasoning**:
I will define the `run_economic_trend_regression` function as requested, constructing the specific regression formula with economic controls and a linear time trend, and configuring it for WLS with clustered standard errors.



In [ ]:
def run_economic_trend_regression(outcome_var, eligibility_var, data):
    """
    Runs a WLS regression of outcome_var on eligibility_var with economic controls,
    linear time trend, and fixed effects.

    Args:
    outcome_var (str): Name of the outcome variable.
    eligibility_var (str): Name of the eligibility variable.
    data (pd.DataFrame): The dataframe containing the data.

    Returns:
    statsmodels.regression.linear_model.RegressionResultsWrapper: The fitted regression results.
    """

    # Construct the formula including economic controls and YEAR as a trend
    # Formula: Outcome ~ Eligibility + Economic_Controls + Linear_Year + Fixed_Effects
    formula = f"{outcome_var} ~ {eligibility_var} + Childhood_Avg_Unemployment + log_childhood_income + YEAR + C(STATE_BPL) + C(AGE) + C(SEX) + C(RACE) + C(HISPAN)"

    # Fit the WLS model
    # We use 'PERWT' as weights (Person Weights)
    model = smf.wls(formula=formula, data=data, weights=data['PERWT'])

    # Fit with clustered standard errors on 'STATE_BPL'
    results = model.fit(cov_type='cluster', cov_kwds={'groups': data['STATE_BPL']})

    return results

print("Function 'run_economic_trend_regression' defined.")

Function 'run_economic_trend_regression' defined.


**Reasoning**:
I will iterate through the combinations of outcomes and eligibility measures, running the newly defined regression function for each. I will collect the key statistics (coefficients, standard errors, p-values) into a list and then create a DataFrame to display the results.



In [ ]:
# Initialize a list to store results
results_economic = []

# Define outcomes and eligibility measures
outcomes = ['hs_grad', 'college_grad', 'years_of_education', 'some_college']
eligibility_measures = ['elig_0_5', 'elig_6_11', 'elig_12_18', 'cumulative_eligibility']

print("Starting economic trend regression analysis...")

# Iterate through each outcome and eligibility measure
for outcome in outcomes:
    for eligibility in eligibility_measures:
        print(f"Running regression for Outcome: {outcome}, Eligibility: {eligibility}...")
        # Run the regression
        results = run_economic_trend_regression(outcome, eligibility, df_clean)

        # Extract results for the eligibility variable and economic controls
        # We need to handle cases where variable names might be slightly different in params
        # usually statsmodels keeps the name unless wrapped in C()

        elig_coef = results.params[eligibility]
        elig_se = results.bse[eligibility]
        elig_p = results.pvalues[eligibility]

        unemp_coef = results.params['Childhood_Avg_Unemployment']
        log_inc_coef = results.params['log_childhood_income']
        year_coef = results.params['YEAR']

        # Append the results
        results_economic.append({
            'Outcome': outcome,
            'Eligibility Variable': eligibility,
            'Eligibility Coef': elig_coef,
            'Eligibility SE': elig_se,
            'Eligibility P-value': elig_p,
            'Unemployment Coef': unemp_coef,
            'Log Income Coef': log_inc_coef,
            'Year Trend Coef': year_coef
        })

# Convert results to a DataFrame
results_economic_df = pd.DataFrame(results_economic)

# Display the results
print("\nEconomic Trend Regression Analysis Results:")
display(results_economic_df)

Starting economic trend regression analysis...
Running regression for Outcome: hs_grad, Eligibility: elig_0_5...
Running regression for Outcome: hs_grad, Eligibility: elig_6_11...
Running regression for Outcome: hs_grad, Eligibility: elig_12_18...
Running regression for Outcome: hs_grad, Eligibility: cumulative_eligibility...
Running regression for Outcome: college_grad, Eligibility: elig_0_5...
Running regression for Outcome: college_grad, Eligibility: elig_6_11...
Running regression for Outcome: college_grad, Eligibility: elig_12_18...
Running regression for Outcome: college_grad, Eligibility: cumulative_eligibility...
Running regression for Outcome: years_of_education, Eligibility: elig_0_5...
Running regression for Outcome: years_of_education, Eligibility: elig_6_11...
Running regression for Outcome: years_of_education, Eligibility: elig_12_18...
Running regression for Outcome: years_of_education, Eligibility: cumulative_eligibility...
Running regression for Outcome: some_college, 

,Outcome,Eligibility Variable,Eligibility Coef,Eligibility SE,Eligibility P-value,Unemployment Coef,Log Income Coef,Year Trend Coef
0,hs_grad,elig_0_5,0.000617,0.007911,0.937783,0.009142,0.009825,0.005306
1,hs_grad,elig_6_11,-0.001656,0.002986,0.579051,0.009486,0.017112,0.005558
2,hs_grad,elig_12_18,-0.001583,0.001557,0.309359,0.008736,0.008644,0.005941
3,hs_grad,cumulative_eligibility,-0.001753,0.001717,0.307491,0.009081,0.014543,0.006316
4,college_grad,elig_0_5,0.002406,0.012440,0.846631,0.002405,0.347468,0.001089
5,college_grad,elig_6_11,-0.000331,0.004772,0.944744,0.002528,0.345198,0.001229
6,college_grad,elig_12_18,0.002064,0.002661,0.437857,0.003012,0.343768,0.000387
7,college_grad,cumulative_eligibility,0.001666,0.002263,0.461666,0.002535,0.338101,0.000247
8,years_of_education,elig_0_5,0.035321,0.062713,0.573290,0.024888,0.845343,0.032397
9,years_of_education,elig_6_11,0.000817,0.029832,0.978165,0.025573,0.783624,0.033668


### Summary of Findings (Economic Trend Specification)

After controlling for childhood economic conditions (unemployment and income) and adding a linear time trend, the regression results indicate the following:

*   **High School Graduation & College Graduation:** There is no statistically significant evidence that Medicaid eligibility in any age window affects the probability of graduating high school or college, as p-values remain well above the 0.05 threshold.
*   **Years of Education:** Similarly, we do not observe a significant impact on total years of education.
*   **Some College:** We observe a **statistically significant negative relationship** between Medicaid eligibility and the likelihood of having "some college" education for the **12-18 age window** (Coef: -0.0053, p < 0.05) and **cumulative eligibility** (Coef: -0.0052, p < 0.05). This suggests that increased eligibility in these later childhood years is associated with a slightly lower probability of attending some college, holding economic factors constant.

Overall, the addition of economic controls and time trends confirms that the link between Medicaid eligibility and educational attainment in this sample is generally weak or non-existent for degree completion, with a small negative association observed for college attendance.

## Final Task

### Subtask:
Summarize the findings from the regression analysis with economic trends.


## Summary:

### Q&A

**Question:** Does Medicaid eligibility significantly impact educational attainment when controlling for economic trends and time?

**Answer:** Generally, no. After controlling for childhood unemployment, log household income, and linear time trends, the analysis found no statistically significant evidence that Medicaid eligibility affects the probability of graduating high school, graduating college, or increasing total years of education. However, a small, statistically significant negative association was found for the "some college" outcome in specific eligibility windows.

### Data Analysis Key Findings

*   **No Significant Impact on Degree Completion:** There were no statistically significant findings linking Medicaid eligibility (across any age window) to High School Graduation, College Graduation, or total Years of Education.
*   **Negative Association with "Some College":** A statistically significant negative relationship was observed between Medicaid eligibility and the likelihood of having "some college" education.
    *   For the **12-18 age window**, the coefficient was **-0.0053** (p < 0.05).
    *   For **cumulative eligibility**, the coefficient was **-0.0052** (p < 0.05).
*   **Economic Controls:** These results hold constant the effects of childhood average unemployment, log childhood income, and a linear year trend (`YEAR`), suggesting the findings are robust to these economic variations.

### Insights or Next Steps

*   **Robustness of Null Results:** The addition of economic controls strengthens the conclusion that Medicaid eligibility is not a primary driver of educational degree completion in this specific sample.
*   **Investigation of "Some College":** The negative correlation in the 12-18 age group for "some college" is counter-intuitive and warrants further investigation to determine if it relates to labor market entry choices or if it acts as a proxy for other unobserved socioeconomic stressors not captured by income/unemployment data.


# Task
Load the dataset `usa_00004.csv` (you can use the existing dataframe `df` if it is already loaded). Create a `birth_year` column by subtracting `AGE` from `YEAR`. Filter the data to keep only individuals born between 1981 and 1984 (inclusive) and who are between the ages of 28 and 34 (inclusive).

Next, verify the data consistency by generating a cross-tabulation of `birth_year` and `AGE`, and print the unique values of `YEAR`.

Then, select the following columns for the final analysis:
*   **Identifiers**: `YEAR`, `SAMPLE`, `SERIAL`, `CBSERIAL`, `HHWT`, `CLUSTER`, `STRATA`, `GQ`
*   **Weights**: `PERWT`
*   **Demographics**: `SEX`, `AGE`, `RACE`, `RACED`, `HISPAN`, `HISPAND`
*   **Geography**: `BPL`, `BPLD`
*   **Education**: `EDUC`, `EDUCD`
*   **Income/Work**: `EMPSTAT`, `EMPSTATD`, `LABFORCE`, `OCC`, `IND`, `WKSWORK2`, `INCTOT`, `INCWAGE`

Rename the column `BPL` (or `STATEFIP_BPL` if that's the name) to `STATE_BPL`. If any columns are missing, print a warning but proceed.

Construct the following outcome variables:
*   `employed`: A binary variable (1 if `EMPSTAT` is 1 or 2, else 0).
*   `log_incwage`: The natural log of `INCWAGE` (replace 0 or negative values with `NaN` before logging).
*   `log_inctot`: The natural log of `INCTOT` (replace 0 or negative values with `NaN` before logging).
*   `weeks_worked_midpoint`: Map `WKSWORK2` to midpoints: {1: 7, 2: 20, 3: 33, 4: 43.5, 5: 48.5, 6: 51}.

Before transforming, print the value counts for `EMPSTAT` and `WKSWORK2`.

Finally, print the shape of the cleaned dataframe, display descriptive statistics for key variables, and check for unique values in `STATE_BPL` and `birth_year`. Crucially, check for missing values in these merge keys (`STATE_BPL`, `birth_year`). Save the cleaned dataframe to `"/content/drive/MyDrive/Colab Notebooks/Thesis/Data/acs_human_capital_clean.csv"` and summarize the processing steps, noting the final sample size.

## Load and Filter Data

### Subtask:
Filter the existing dataframe `df` for the specified birth cohorts (1981-1984) and age range (28-34).


## Clean Data and Re-run Regressions

### Subtask:
Drop rows with missing values and re-run the WLS regression analysis for all combinations of outcomes and eligibility measures using the clean dataset.


In [3]:
import pandas as pd

file_path = '/content/drive/MyDrive/Colab Notebooks/Thesis/Data/Mechanisms Data /(Human Capital) usa_00004.csv'
# Load into a specific dataframe to avoid overlap
df_acs = pd.read_csv(file_path)

# Create 'birth_year' variable
df_acs['birth_year'] = df_acs['YEAR'] - df_acs['AGE']

# Filter for ages 28-34
df_acs = df_acs[(df_acs['AGE'] >= 28) & (df_acs['AGE'] <= 34)]

# Filter for birth years 1981-1984
df_acs = df_acs[(df_acs['birth_year'] >= 1981) & (df_acs['birth_year'] <= 1984)]

# Display the first few rows of the filtered dataframe
display(df_acs.head())

,YEAR,SAMPLE,SERIAL,CBSERIAL,HHWT,CLUSTER,STATEFIP,STRATA,GQ,PERNUM,...,EDUCD,EMPSTAT,EMPSTATD,CLASSWKR,CLASSWKRD,WKSWORK2,UHRSWORK,INCTOT,INCWAGE,birth_year
3000740,2009,200901,42,2992,126.0,2009000000421,1,70001,1,3,...,63,1,10,2,22,6,40,26500,26500,1981
3000753,2009,200901,48,3322,110.0,2009000000481,1,40001,1,1,...,63,1,10,2,22,6,40,38100,32100,1981
3000754,2009,200901,48,3322,110.0,2009000000481,1,40001,1,2,...,63,1,10,2,22,6,40,22800,22800,1981
3000799,2009,200901,68,4312,263.0,2009000000681,1,90101,1,1,...,63,1,10,2,22,6,40,26000,26000,1981
3000887,2009,200901,107,6937,27.0,2009000001071,1,40001,1,2,...,81,1,10,2,22,6,40,30000,30000,1981


In [4]:
print(f"Number of rows in filtered dataset: {len(df_acs)}")

Number of rows in filtered dataset: 1036357


# Task
Check data consistency for `df_acs` by generating a cross-tabulation of `birth_year` and `AGE`, and print unique `YEAR` values.

Then, process the education variables in `df_acs`:
1.  Drop `STATEFIP` if it exists.
2.  Rename `BPL` to `STATE_BPL`.
3.  Remove rows where `EDUCD` is 0, 1, or 999.
4.  Create binary variables:
    *   `hs_grad`: 1 if `EDUCD` is between 62 and 116 (inclusive), else 0.
    *   `college_grad`: 1 if `EDUCD` is between 101 and 116 (inclusive), else 0.
    *   `some_college`: 1 if `EDUCD` is between 65 and 116 (inclusive), else 0.
5.  Create `years_of_education` using the specific mapping logic provided earlier (e.g., `educd <= 2 -> 0`, `educd == 116 -> 20`, etc.).

Next, process the economic variables:
1.  Print value counts for `EMPSTAT` and `WKSWORK2` to inspect the raw data.
2.  Create `employed`: 1 if `EMPSTAT` equals 1 (Employed), else 0.
3.  Create `log_incwage`: The natural log of `INCWAGE` (replace 0 or negative values with `NaN` first).
4.  Create `log_inctot`: The natural log of `INCTOT` (replace 0 or negative values with `NaN` first).
5.  Create `weeks_worked_midpoint` by mapping `WKSWORK2` values: `{1: 7, 2: 20, 3: 33, 4: 43.5, 5: 48.5, 6: 51}`.
6.  Ensure the column `UHRSWORK` (hours worked per week) is kept if available (or `HRSWORK2` if `UHRSWORK` is missing, check availability).

Finally, print the dataframe shape, descriptive statistics for the newly created variables, and check for missing values in the merge keys (`STATE_BPL`, `birth_year`). Save the cleaned dataframe to `"/content/drive/MyDrive/Colab Notebooks/Thesis/Data/acs_human_capital_clean.csv"` and summarize the cleaning process.

## Verify Data Consistency

### Subtask:
Check the consistency of birth year and age calculations and inspect unique survey years.


**Reasoning**:
Perform a cross-tabulation of birth year and age to verify cohort consistency and check the unique survey years present in the filtered dataset.



In [5]:
# Cross-tabulation of birth_year and AGE
print("Cross-tabulation of Birth Year and Age:")
display(pd.crosstab(df_acs['birth_year'], df_acs['AGE']))

# Unique values of YEAR
print("\nUnique Survey Years:")
print(df_acs['YEAR'].unique())

Cross-tabulation of Birth Year and Age:


AGE,28,29,30,31,32,33,34
birth_year,,,,,,,
1981,35440,36183,37688,36318,37814,37276,37578
1982,36155,35960,37522,36675,37192,37193,37482
1983,36273,35534,38325,36431,37340,37779,39083
1984,35170,35840,37909,36507,37229,38071,38390



Unique Survey Years:
[2009 2010 2011 2012 2013 2014 2015 2016 2017 2018]


**Reasoning**:
I will process the education and economic variables as requested. This involves renaming columns, filtering invalid education codes, creating binary education outcomes, mapping years of education, creating employment and income variables (including log transformations), and mapping weeks worked. Finally, I will check for missing values in key columns and save the cleaned dataset.



In [6]:
import numpy as np

# --- Education Variables ---

# Drop STATEFIP if it exists
if 'STATEFIP' in df_acs.columns:
    df_acs = df_acs.drop(columns=['STATEFIP'])

# Rename BPL to STATE_BPL
if 'BPL' in df_acs.columns:
    df_acs = df_acs.rename(columns={'BPL': 'STATE_BPL'})

# Remove rows where EDUCD is 0, 1, or 999
df_acs = df_acs[~df_acs['EDUCD'].isin([0, 1, 999])].copy()

# Create binary variables
df_acs['hs_grad'] = ((df_acs['EDUCD'] >= 62) & (df_acs['EDUCD'] <= 116)).astype(int)
df_acs['college_grad'] = ((df_acs['EDUCD'] >= 101) & (df_acs['EDUCD'] <= 116)).astype(int)
df_acs['some_college'] = ((df_acs['EDUCD'] >= 65) & (df_acs['EDUCD'] <= 116)).astype(int)

# Years of Education Mapping
def map_educd_to_years(educd):
    if educd <= 2: return 0
    elif educd == 10: return 2.5
    elif educd == 14: return 1
    elif educd == 15: return 2
    elif educd == 16: return 3
    elif educd == 17: return 4
    elif educd == 21: return 5.5
    elif educd == 22: return 5
    elif educd == 23: return 6
    elif educd == 24: return 7.5
    elif educd == 25: return 7
    elif educd == 26: return 8
    elif educd == 30: return 9
    elif educd == 40: return 10
    elif educd == 50: return 11
    elif educd in [60, 61, 62, 63, 64]: return 12
    elif educd == 65: return 12.5
    elif educd in [70, 71]: return 13
    elif educd in [80, 81, 82, 83]: return 14
    elif educd == 90: return 15
    elif educd in [100, 101]: return 16
    elif educd == 110: return 17
    elif educd == 111: return 18
    elif educd == 112: return 19
    elif educd == 113: return 20
    elif educd == 114: return 18
    elif educd == 115: return 19
    elif educd == 116: return 20
    else: return np.nan

df_acs['years_of_education'] = df_acs['EDUCD'].apply(map_educd_to_years)

# --- Economic Variables ---

# Print initial value counts
print("Original EMPSTAT value counts:")
print(df_acs['EMPSTAT'].value_counts())
print("\nOriginal WKSWORK2 value counts:")
print(df_acs['WKSWORK2'].value_counts())

# Create employed variable (1 if EMPSTAT is 1, else 0)
# Note: Task description said "1 if EMPSTAT is 1 or 2", but later instruction said "1 if EMPSTAT equals 1".
# Usually EMPSTAT 1=Employed, 2=Unemployed, 3=Not in Labor Force.
# I will follow the explicit instruction "1 if EMPSTAT equals 1 (Employed)" from step text.
df_acs['employed'] = (df_acs['EMPSTAT'] == 1).astype(int)

# Create log income variables (replace <= 0 with NaN first)
df_acs['log_incwage'] = np.log(df_acs['INCWAGE'].where(df_acs['INCWAGE'] > 0))
df_acs['log_inctot'] = np.log(df_acs['INCTOT'].where(df_acs['INCTOT'] > 0))

# Create weeks_worked_midpoint
wkswork2_map = {1: 7, 2: 20, 3: 33, 4: 43.5, 5: 48.5, 6: 51}
df_acs['weeks_worked_midpoint'] = df_acs['WKSWORK2'].map(wkswork2_map)

# Check/Keep UHRSWORK or HRSWORK2
if 'UHRSWORK' in df_acs.columns:
    print("\nUsing UHRSWORK for hours worked.")
elif 'HRSWORK2' in df_acs.columns:
    print("\nUHRSWORK not found, using HRSWORK2.")
else:
    print("\nNeither UHRSWORK nor HRSWORK2 found.")

# --- Final Checks & Save ---

print(f"\nShape of cleaned dataframe: {df_acs.shape}")

# Check for missing values in merge keys
print("\nMissing values in merge keys:")
print(df_acs[['STATE_BPL', 'birth_year']].isnull().sum())

# Descriptive statistics for new variables
new_vars = ['hs_grad', 'college_grad', 'years_of_education', 'employed', 'log_incwage', 'weeks_worked_midpoint']
print("\nDescriptive Statistics for new variables:")
display(df_acs[new_vars].describe())

# Save to CSV
output_path_clean = '/content/drive/MyDrive/Colab Notebooks/Thesis/Data/acs_human_capital_clean.csv'
df_acs.to_csv(output_path_clean, index=False)
print(f"\nCleaned dataframe saved to: {output_path_clean}")

Original EMPSTAT value counts:
EMPSTAT
1    777002
3    198859
2     60496
Name: count, dtype: int64

Original WKSWORK2 value counts:
WKSWORK2
6    658671
0    185539
4     50507
3     48437
1     40273
2     34805
5     18125
Name: count, dtype: int64

Using UHRSWORK for hours worked.

Shape of cleaned dataframe: (1036357, 37)

Missing values in merge keys:
STATE_BPL     0
birth_year    0
dtype: int64

Descriptive Statistics for new variables:


,hs_grad,college_grad,years_of_education,employed,log_incwage,weeks_worked_midpoint
count,1.036357e+06,1.036357e+06,1.035984e+06,1.036357e+06,815821.000000,850818.000000
mean,8.927358e-01,3.619380e-01,1.375516e+01,7.497436e-01,10.230012,46.125928
std,3.094490e-01,4.805612e-01,2.986333e+00,4.331609e-01,1.057978,11.362788
min,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,1.386294,7.000000
25%,1.000000e+00,0.000000e+00,1.200000e+01,0.000000e+00,9.809177,51.000000
50%,1.000000e+00,0.000000e+00,1.300000e+01,1.000000e+00,10.431170,51.000000
75%,1.000000e+00,1.000000e+00,1.600000e+01,1.000000e+00,10.878047,51.000000
max,1.000000e+00,1.000000e+00,2.000000e+01,1.000000e+00,13.508985,51.000000



Cleaned dataframe saved to: /content/drive/MyDrive/Colab Notebooks/Thesis/Data/acs_human_capital_clean.csv


## Final Task

### Subtask:
Summarize the cleaning process and the final dataset properties.


## Summary:

### Q&A

**Question:** What are the unique survey years and does the age/birth year data appear consistent?

**Answer:** The survey data spans the years **2009 to 2018**. The cross-tabulation of `birth_year` and `AGE` confirmed logical consistency, with specific cohorts (such as those born in 1981) displaying the appropriate age progression (ages 28-34) across the observed survey window.

### Data Analysis Key Findings

*   **Final Dataset Dimensions:** After filtering out invalid education codes (0, 1, 999) and generating new features, the final cleaned dataframe contains **1,036,357 rows** and **37 columns**.
*   **Data Integrity:** The cleaning process confirmed there are **0 missing values** in the primary merge keys, `STATE_BPL` and `birth_year`.
*   **Feature Engineering:** Multiple derived variables were successfully constructed:
    *   **Education:** Binary flags for high school and college graduation, plus a continuous `years_of_education` variable mapped from standard codes.
    *   **Economics:** Logarithmic transformations of income (`log_incwage`, `log_inctot`) and a standardized `weeks_worked_midpoint` based on interval codes.

### Insights or Next Steps

*   The dataset is now strictly standardized and free of missing keys, making it ready for immediate merging with other state-level or cohort-level datasets using `STATE_BPL` and `birth_year`.
*   The creation of logarithmic income variables suggests the analysis is prepared for regression modeling where wage elasticity or percentage changes are the focus.


# Task
Load the file `"[Final]_US_Crime_Rates_1999_2010.csv"` from `/content/drive/MyDrive/Colab Notebooks/Thesis/Data/`. Select the columns `fips`, `brthyr`, `cumulative_eligibility`, `elig_0_5`, `elig_6_11`, `elig_12_18`, `Childhood_Avg_Unemployment`, and `childhood_avg_income`. Rename `fips` to `STATE_BPL` and `brthyr` to `birth_year`. Drop duplicates based on `STATE_BPL` and `birth_year` to create a unique lookup table for control variables.

Merge this control table into the existing `df_acs` dataframe using a left join on `STATE_BPL` and `birth_year`. After merging, create a new column `log_childhood_income` by taking the natural log of `childhood_avg_income`. Check for missing values in the newly added columns to verify the merge quality. Finally, save the fully enriched dataframe to `"/content/drive/MyDrive/Colab Notebooks/Thesis/Data/acs_human_capital_merged.csv"` and summarize the merging process.

## Load External Data

### Subtask:
Load the US Crime Rates CSV file and inspect it.


**Reasoning**:
Load the US Crime Rates CSV file and inspect its contents as per the instructions.



In [7]:
file_path_crime = '/content/drive/MyDrive/Colab Notebooks/Thesis/Data/[Final]_US_Crime_Rates_1999_2010.csv'
df_crime_rates = pd.read_csv(file_path_crime)

# Display the first few rows of the dataframe
display(df_crime_rates.head())

# Print the column names
print(df_crime_rates.columns)

,fips,year,age,crime_count,brthyr,Age,State,Population,Drug_Rate,Financial_Rate,...,Violent_Rate,cumulative_eligibility,elig_0_5,elig_6_11,elig_12_18,trend_cohort,Childhood_Avg_Unemployment,economically_motivated_crimes,state_abb,childhood_avg_income
0,1,2000,19,1594,1981,19,AL,67769,142.395491,57.400877,...,168.661187,3.228658,1.000340,0.858396,1.369921,0,7.957895,235.210790,AL,10537
1,1,2001,19,1741,1982,19,AL,65031,141.009672,66.122311,...,182.374560,3.559218,0.963729,0.939953,1.655536,1,7.642105,267.718473,AL,10732
2,1,2001,20,1686,1981,20,AL,68511,133.263272,87.285253,...,178.949366,3.228658,1.000340,0.858396,1.369921,0,7.957895,246.091868,AL,10537
3,1,2002,19,1758,1983,19,AL,64169,154.124266,55.946018,...,213.342891,3.929335,0.927237,0.999889,2.002209,2,7.168421,273.964064,AL,10937
4,1,2002,20,1655,1982,20,AL,65515,152.636801,69.449744,...,216.896894,3.559218,0.963729,0.939953,1.655536,1,7.642105,252.613905,AL,10732


Index(['fips', 'year', 'age', 'crime_count', 'brthyr', 'Age', 'State',
       'Population', 'Drug_Rate', 'Financial_Rate', 'Property_Rate',
       'PublicOrder_Rate', 'Violent_Rate', 'cumulative_eligibility',
       'elig_0_5', 'elig_6_11', 'elig_12_18', 'trend_cohort',
       'Childhood_Avg_Unemployment', 'economically_motivated_crimes',
       'state_abb', 'childhood_avg_income'],
      dtype='object')


**Reasoning**:
I will select the specified control variables from the crime rates dataframe, rename the identifier columns to match the ACS dataset, and remove duplicates to create a unique lookup table. Then, I will merge this table into the ACS dataframe, create the logarithmic income variable, check for missing values, and save the final dataset.



In [8]:
import numpy as np

# 1. Select relevant columns from crime rates data
control_cols = [
    'fips', 'brthyr',
    'cumulative_eligibility', 'elig_0_5', 'elig_6_11', 'elig_12_18',
    'Childhood_Avg_Unemployment', 'childhood_avg_income'
]
df_controls = df_crime_rates[control_cols].copy()

# 2. Rename columns to match df_acs
df_controls = df_controls.rename(columns={'fips': 'STATE_BPL', 'brthyr': 'birth_year'})

# 3. Drop duplicates to create a unique lookup table
# We take the first occurrence as these values are constant for a State-Cohort pair
df_controls_unique = df_controls.drop_duplicates(subset=['STATE_BPL', 'birth_year'])

# 4. Merge into df_acs
# Ensure merge keys are the same type (int)
df_acs['STATE_BPL'] = df_acs['STATE_BPL'].astype(int)
df_acs['birth_year'] = df_acs['birth_year'].astype(int)
df_controls_unique['STATE_BPL'] = df_controls_unique['STATE_BPL'].astype(int)
df_controls_unique['birth_year'] = df_controls_unique['birth_year'].astype(int)

df_acs_merged = pd.merge(df_acs, df_controls_unique, on=['STATE_BPL', 'birth_year'], how='left')

# 5. Create log_childhood_income
# Replace <= 0 with NaN before log
df_acs_merged['log_childhood_income'] = np.log(
    df_acs_merged['childhood_avg_income'].where(df_acs_merged['childhood_avg_income'] > 0)
)

# 6. Check for missing values in new columns
new_cols = ['cumulative_eligibility', 'Childhood_Avg_Unemployment', 'log_childhood_income']
print("Missing values in merged control variables:")
print(df_acs_merged[new_cols].isnull().sum())

# 7. Save the final merged dataframe
output_path_merged = '/content/drive/MyDrive/Colab Notebooks/Thesis/Data/acs_human_capital_merged.csv'
df_acs_merged.to_csv(output_path_merged, index=False)

print(f"\nFinal merged dataframe shape: {df_acs_merged.shape}")
print(f"Data saved to: {output_path_merged}")

# Display first few rows
display(df_acs_merged.head())

/tmp/ipykernel_5554/321468139.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_controls_unique['STATE_BPL'] = df_controls_unique['STATE_BPL'].astype(int)
/tmp/ipykernel_5554/321468139.py:23: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_controls_unique['birth_year'] = df_controls_unique['birth_year'].astype(int)


Missing values in merged control variables:
cumulative_eligibility        219622
Childhood_Avg_Unemployment    219622
log_childhood_income          219622
dtype: int64

Final merged dataframe shape: (1036357, 44)
Data saved to: /content/drive/MyDrive/Colab Notebooks/Thesis/Data/acs_human_capital_merged.csv


,YEAR,SAMPLE,SERIAL,CBSERIAL,HHWT,CLUSTER,STRATA,GQ,PERNUM,PERWT,...,log_incwage,log_inctot,weeks_worked_midpoint,cumulative_eligibility,elig_0_5,elig_6_11,elig_12_18,Childhood_Avg_Unemployment,childhood_avg_income,log_childhood_income
0,2009,200901,42,2992,126.0,2009000000421,70001,1,3,94.0,...,10.184900,10.184900,51.0,1.828912,0.650532,0.612582,0.565799,8.189474,9177.0,9.124456
1,2009,200901,48,3322,110.0,2009000000481,40001,1,1,111.0,...,10.376611,10.547970,51.0,3.252349,0.905107,0.649300,1.697942,5.705263,11898.0,9.384126
2,2009,200901,48,3322,110.0,2009000000481,40001,1,2,110.0,...,10.034516,10.034516,51.0,3.142020,0.998988,0.945851,1.197181,6.610526,12032.0,9.395325
3,2009,200901,68,4312,263.0,2009000000681,90101,1,1,263.0,...,10.165852,10.165852,51.0,3.228658,1.000340,0.858396,1.369921,7.957895,10537.0,9.262648
4,2009,200901,107,6937,27.0,2009000001071,40001,1,2,33.0,...,10.308953,10.308953,51.0,3.228658,1.000340,0.858396,1.369921,7.957895,10537.0,9.262648


**Reasoning**:
The previous execution triggered a `SettingWithCopyWarning` because `df_controls_unique` was created as a slice (view) of `df_controls` without an explicit copy before assignment. I will fix this by appending `.copy()` after `drop_duplicates()`. I will also use `.loc` or explicit copying when modifying `df_acs` columns to ensure safe assignment.



In [9]:
import numpy as np
import pandas as pd

# 1. Select relevant columns from crime rates data
control_cols = [
    'fips', 'brthyr',
    'cumulative_eligibility', 'elig_0_5', 'elig_6_11', 'elig_12_18',
    'Childhood_Avg_Unemployment', 'childhood_avg_income'
]
df_controls = df_crime_rates[control_cols].copy()

# 2. Rename columns to match df_acs
df_controls = df_controls.rename(columns={'fips': 'STATE_BPL', 'brthyr': 'birth_year'})

# 3. Drop duplicates to create a unique lookup table
# Use .copy() to avoid SettingWithCopyWarning
df_controls_unique = df_controls.drop_duplicates(subset=['STATE_BPL', 'birth_year']).copy()

# 4. Merge into df_acs
# Ensure merge keys are the same type (int)
# Use .loc to avoid SettingWithCopyWarning on df_acs if it's a view
df_acs.loc[:, 'STATE_BPL'] = df_acs['STATE_BPL'].astype(int)
df_acs.loc[:, 'birth_year'] = df_acs['birth_year'].astype(int)

df_controls_unique['STATE_BPL'] = df_controls_unique['STATE_BPL'].astype(int)
df_controls_unique['birth_year'] = df_controls_unique['birth_year'].astype(int)

df_acs_merged = pd.merge(df_acs, df_controls_unique, on=['STATE_BPL', 'birth_year'], how='left')

# 5. Create log_childhood_income
# Replace <= 0 with NaN before log
df_acs_merged['log_childhood_income'] = np.log(
    df_acs_merged['childhood_avg_income'].where(df_acs_merged['childhood_avg_income'] > 0)
)

# 6. Check for missing values in new columns
new_cols = ['cumulative_eligibility', 'Childhood_Avg_Unemployment', 'log_childhood_income']
print("Missing values in merged control variables:")
print(df_acs_merged[new_cols].isnull().sum())

# 7. Save the final merged dataframe
output_path_merged = '/content/drive/MyDrive/Colab Notebooks/Thesis/Data/acs_human_capital_merged.csv'
df_acs_merged.to_csv(output_path_merged, index=False)

print(f"\nFinal merged dataframe shape: {df_acs_merged.shape}")
print(f"Data saved to: {output_path_merged}")

# Display first few rows
display(df_acs_merged.head())

Missing values in merged control variables:
cumulative_eligibility        219622
Childhood_Avg_Unemployment    219622
log_childhood_income          219622
dtype: int64

Final merged dataframe shape: (1036357, 44)
Data saved to: /content/drive/MyDrive/Colab Notebooks/Thesis/Data/acs_human_capital_merged.csv


,YEAR,SAMPLE,SERIAL,CBSERIAL,HHWT,CLUSTER,STRATA,GQ,PERNUM,PERWT,...,log_incwage,log_inctot,weeks_worked_midpoint,cumulative_eligibility,elig_0_5,elig_6_11,elig_12_18,Childhood_Avg_Unemployment,childhood_avg_income,log_childhood_income
0,2009,200901,42,2992,126.0,2009000000421,70001,1,3,94.0,...,10.184900,10.184900,51.0,1.828912,0.650532,0.612582,0.565799,8.189474,9177.0,9.124456
1,2009,200901,48,3322,110.0,2009000000481,40001,1,1,111.0,...,10.376611,10.547970,51.0,3.252349,0.905107,0.649300,1.697942,5.705263,11898.0,9.384126
2,2009,200901,48,3322,110.0,2009000000481,40001,1,2,110.0,...,10.034516,10.034516,51.0,3.142020,0.998988,0.945851,1.197181,6.610526,12032.0,9.395325
3,2009,200901,68,4312,263.0,2009000000681,90101,1,1,263.0,...,10.165852,10.165852,51.0,3.228658,1.000340,0.858396,1.369921,7.957895,10537.0,9.262648
4,2009,200901,107,6937,27.0,2009000001071,40001,1,2,33.0,...,10.308953,10.308953,51.0,3.228658,1.000340,0.858396,1.369921,7.957895,10537.0,9.262648


In [10]:
# 1. Check if STATE_BPL itself is missing
print("Missing STATE_BPL values:", df_acs_merged['STATE_BPL'].isnull().sum())

# 2. Identify which STATE_BPL values have missing cumulative_eligibility
missing_eligibility = df_acs_merged[df_acs_merged['cumulative_eligibility'].isnull()]
unique_states_missing = missing_eligibility['STATE_BPL'].unique()
print(f"\nNumber of rows with missing eligibility: {len(missing_eligibility)}")
print(f"States (BPL codes) with missing eligibility data: {sorted(unique_states_missing)}")

# 3. Drop rows where cumulative_eligibility is NaN
# This effectively drops rows where the merge failed due to missing control data
df_acs_clean_merged = df_acs_merged.dropna(subset=['cumulative_eligibility']).copy()

# 4. Verify the cleanup
print(f"\nRows before dropping: {len(df_acs_merged)}")
print(f"Rows after dropping: {len(df_acs_clean_merged)}")
print(f"Rows dropped: {len(df_acs_merged) - len(df_acs_clean_merged)}")

# 5. Check for any remaining missing values in control columns
control_cols_check = ['cumulative_eligibility', 'Childhood_Avg_Unemployment', 'log_childhood_income']
print("\nRemaining missing values in control columns:")
print(df_acs_clean_merged[control_cols_check].isnull().sum())

Missing STATE_BPL values: 0

Number of rows with missing eligibility: 219622
States (BPL codes) with missing eligibility data: [np.int64(12), np.int64(100), np.int64(105), np.int64(110), np.int64(115), np.int64(150), np.int64(160), np.int64(200), np.int64(210), np.int64(250), np.int64(260), np.int64(299), np.int64(300), np.int64(400), np.int64(401), np.int64(402), np.int64(404), np.int64(405), np.int64(410), np.int64(411), np.int64(413), np.int64(414), np.int64(420), np.int64(421), np.int64(425), np.int64(426), np.int64(430), np.int64(433), np.int64(434), np.int64(436), np.int64(438), np.int64(450), np.int64(451), np.int64(452), np.int64(453), np.int64(454), np.int64(455), np.int64(456), np.int64(457), np.int64(460), np.int64(461), np.int64(462), np.int64(465), np.int64(499), np.int64(500), np.int64(501), np.int64(502), np.int64(509), np.int64(511), np.int64(512), np.int64(513), np.int64(514), np.int64(515), np.int64(516), np.int64(517), np.int64(518), np.int64(520), np.int64(521), np.

In [11]:
# Save the cleaned, merged dataframe
output_path_final = '/content/drive/MyDrive/Colab Notebooks/Thesis/Data/acs_human_capital_merged.csv'
df_acs_clean_merged.to_csv(output_path_final, index=False)

print(f"Final cleaned dataset saved to: {output_path_final}")
print(f"Final row count: {len(df_acs_clean_merged)}")

Final cleaned dataset saved to: /content/drive/MyDrive/Colab Notebooks/Thesis/Data/acs_human_capital_merged.csv
Final row count: 816735


# Task
Run the WLS regression analysis using the newly created dataframe `df_acs_clean_merged`.

First, define the outcomes: `employed`, `log_incwage`, `log_inctot`, and `weeks_worked_midpoint`.
The eligibility measures remain: `elig_0_5`, `elig_6_11`, `elig_12_18`, and `cumulative_eligibility`.

Iterate through each combination of outcome and eligibility measure. For each combination, run the `run_economic_trend_regression` function (defined previously) using `df_acs_clean_merged`. This function controls for `Childhood_Avg_Unemployment`, `log_childhood_income`, `YEAR`, `STATE_BPL` (fixed effects), `AGE`, `SEX`, `RACE`, and `HISPAN`.

Collect the results (Outcome, Eligibility Variable, Eligibility Coef, Eligibility SE, Eligibility P-value, Unemployment Coef, Log Income Coef, Year Trend Coef) into a list, convert it to a pandas DataFrame, and display the final results table. Summarize the findings, specifically looking for any significant impacts of Medicaid eligibility on these labor market outcomes.

## Save Final Dataset

### Subtask:
Save the dataframe `df_acs_clean_merged` to `/content/drive/MyDrive/Colab Notebooks/Thesis/Data/acs_human_capital_final.csv`.


**Reasoning**:
Save the final dataframe to the specified CSV file path as requested.



In [12]:
output_path_final_dataset = '/content/drive/MyDrive/Colab Notebooks/Thesis/Data/acs_human_capital_final.csv'
df_acs_clean_merged.to_csv(output_path_final_dataset, index=False)

print(f"Final dataset successfully saved to: {output_path_final_dataset}")

Final dataset successfully saved to: /content/drive/MyDrive/Colab Notebooks/Thesis/Data/acs_human_capital_final.csv


# Task
Run WLS regression analysis using `df_acs_clean_merged` to evaluate the impact of Medicaid eligibility on four labor market outcomes (`employed`, `log_incwage`, `log_inctot`, `weeks_worked_midpoint`) across four eligibility measures (`elig_0_5`, `elig_6_11`, `elig_12_18`, `cumulative_eligibility`). The regression will include controls for `Childhood_Avg_Unemployment`, `log_childhood_income`, `YEAR`, and fixed effects for `STATE_BPL`, `AGE`, `SEX`, `RACE`, and `HISPAN`. Display the compiled results (coefficients, standard errors, and p-values for key variables) in a DataFrame and provide a brief summary of the findings.

## Import Libraries and Define Regression Function

### Subtask:
Import the necessary statsmodels libraries and define a function to perform WLS regression with clustered standard errors for the specified model.


**Reasoning**:
I will define the `run_wls_regression` function as specified, which constructs the regression formula including all controls, fits a WLS model with `PERWT` as weights, calculates clustered standard errors by `STATE_BPL`, and returns the requested statistics (coefficient, standard error, p-value for the eligibility variable, R-squared, and number of observations).



In [13]:
def run_wls_regression(outcome_var, eligibility_var, data):
    """
    Runs a WLS regression of outcome_var on eligibility_var with economic controls,
    linear time trend, and fixed effects, returning specific statistics.

    Args:
    outcome_var (str): Name of the outcome variable.
    eligibility_var (str): Name of the eligibility variable.
    data (pd.DataFrame): The dataframe containing the data.

    Returns:
    tuple: (elig_coef, elig_se, elig_pvalue, rsquared, nobs) for the eligibility variable and model.
    """

    # Construct the formula including economic controls and YEAR as a trend
    # Formula: Outcome ~ Eligibility + Economic_Controls + Linear_Year + Fixed_Effects
    formula = f"{outcome_var} ~ {eligibility_var} + Childhood_Avg_Unemployment + log_childhood_income + YEAR + C(STATE_BPL) + C(AGE) + C(SEX) + C(RACE) + C(HISPAN)"

    # Fit the WLS model
    # We use 'PERWT' as weights (Person Weights)
    model = smf.wls(formula=formula, data=data, weights=data['PERWT'])

    # Fit with clustered standard errors on 'STATE_BPL'
    results = model.fit(cov_type='cluster', cov_kwds={'groups': data['STATE_BPL']})

    # Extract results for the eligibility variable
    elig_coef = results.params[eligibility_var]
    elig_se = results.bse[eligibility_var]
    elig_pvalue = results.pvalues[eligibility_var]

    # Extract R-squared and number of observations
    rsquared = results.rsquared
    nobs = results.nobs

    return elig_coef, elig_se, elig_pvalue, rsquared, nobs

print("Function 'run_wls_regression' defined.")

Function 'run_wls_regression' defined.


**Reasoning**:
I will run the WLS regression analysis using the `df_acs_clean_merged` dataframe. I will define the specified outcomes and eligibility measures, iterate through their combinations, and use the `run_wls_regression` function to gather the coefficients, standard errors, p-values, R-squared, and number of observations for each regression. Finally, I will compile these results into a pandas DataFrame and display it.



In [16]:
def run_wls_regression(outcome_var, eligibility_var, data):
    """
    Runs a WLS regression of outcome_var on eligibility_var with economic controls,
    linear time trend, and fixed effects, returning specific statistics.

    Args:
    outcome_var (str): Name of the outcome variable.
    eligibility_var (str): Name of the eligibility variable.
    data (pd.DataFrame): The dataframe containing the data.

    Returns:
    tuple: (elig_coef, elig_se, elig_pvalue, rsquared, nobs) for the eligibility variable and model.
    """

    # Define all variables involved in the regression (outcome, eligibility, controls, weights, clustering groups)
    all_vars_for_model = [
        outcome_var, eligibility_var, 'Childhood_Avg_Unemployment',
        'log_childhood_income', 'YEAR', 'STATE_BPL', 'AGE', 'SEX',
        'RACE', 'HISPAN', 'PERWT' # PERWT for weights, STATE_BPL for clustering
    ]

    # Create a temporary dataframe with only the necessary columns and drop rows with any NaNs in these columns
    # This ensures alignment between data, weights, and clustering groups
    temp_data = data[all_vars_for_model].dropna().copy()

    # Construct the formula including economic controls and YEAR as a trend
    # Formula: Outcome ~ Eligibility + Economic_Controls + Linear_Year + Fixed_Effects
    formula = f"{outcome_var} ~ {eligibility_var} + Childhood_Avg_Unemployment + log_childhood_income + YEAR + C(STATE_BPL) + C(AGE) + C(SEX) + C(RACE) + C(HISPAN)"

    # Fit the WLS model
    # We use 'PERWT' as weights (Person Weights)
    model = smf.wls(formula=formula, data=temp_data, weights=temp_data['PERWT'])

    # Fit with clustered standard errors on 'STATE_BPL'
    # The groups should come from the same temp_data to ensure alignment
    results = model.fit(cov_type='cluster', cov_kwds={'groups': temp_data['STATE_BPL']})

    # Extract results for the eligibility variable
    elig_coef = results.params[eligibility_var]
    elig_se = results.bse[eligibility_var]
    elig_pvalue = results.pvalues[eligibility_var]

    # Extract R-squared and number of observations
    rsquared = results.rsquared
    nobs = results.nobs

    return elig_coef, elig_se, elig_pvalue, rsquared, nobs

print("Function 'run_wls_regression' defined.")

Function 'run_wls_regression' defined.


**Reasoning**:
I will re-run the WLS regression analysis using the `df_acs_clean_merged` dataframe and the newly defined `run_wls_regression` function. This will iterate through the specified outcomes and eligibility measures, gather the coefficients, standard errors, p-values, R-squared, and number of observations for each regression, and then display the results in a pandas DataFrame.



In [17]:
import pandas as pd
import statsmodels.formula.api as smf
import statsmodels.api as sm

# Define outcomes and eligibility measures
outcomes = ['employed', 'log_incwage', 'log_inctot', 'weeks_worked_midpoint']
eligibility_measures = ['elig_0_5', 'elig_6_11', 'elig_12_18', 'cumulative_eligibility']

# Initialize a list to store results
results_labor_market = []

print("Starting labor market regression analysis...")

# Iterate through each outcome and eligibility measure
for outcome in outcomes:
    for eligibility in eligibility_measures:
        print(f"Running regression for Outcome: {outcome}, Eligibility: {eligibility}...")
        # Run the regression
        elig_coef, elig_se, elig_pvalue, rsquared, nobs = run_wls_regression(outcome, eligibility, df_acs_clean_merged)

        # Append the results
        results_labor_market.append({
            'Outcome': outcome,
            'Eligibility Variable': eligibility,
            'Eligibility Coef': elig_coef,
            'Eligibility SE': elig_se,
            'Eligibility P-value': elig_pvalue,
            'R-squared': rsquared,
            'Nobs': nobs
        })

# Convert results to a DataFrame
results_labor_market_df = pd.DataFrame(results_labor_market)

# Display the results
print("\nLabor Market Regression Analysis Results:")
display(results_labor_market_df)

Starting labor market regression analysis...
Running regression for Outcome: employed, Eligibility: elig_0_5...
Running regression for Outcome: employed, Eligibility: elig_6_11...
Running regression for Outcome: employed, Eligibility: elig_12_18...
Running regression for Outcome: employed, Eligibility: cumulative_eligibility...
Running regression for Outcome: log_incwage, Eligibility: elig_0_5...
Running regression for Outcome: log_incwage, Eligibility: elig_6_11...
Running regression for Outcome: log_incwage, Eligibility: elig_12_18...
Running regression for Outcome: log_incwage, Eligibility: cumulative_eligibility...
Running regression for Outcome: log_inctot, Eligibility: elig_0_5...
Running regression for Outcome: log_inctot, Eligibility: elig_6_11...
Running regression for Outcome: log_inctot, Eligibility: elig_12_18...
Running regression for Outcome: log_inctot, Eligibility: cumulative_eligibility...
Running regression for Outcome: weeks_worked_midpoint, Eligibility: elig_0_5...


,Outcome,Eligibility Variable,Eligibility Coef,Eligibility SE,Eligibility P-value,R-squared,Nobs
0,employed,elig_0_5,0.004909,0.010685,0.645934,0.023323,816735.0
1,employed,elig_6_11,-0.000299,0.005437,0.956103,0.023322,816735.0
2,employed,elig_12_18,0.006522,0.002732,0.016961,0.023334,816735.0
3,employed,cumulative_eligibility,0.005381,0.002439,0.027373,0.023332,816735.0
4,log_incwage,elig_0_5,-0.019939,0.024143,0.408879,0.058913,652475.0
5,log_incwage,elig_6_11,0.018210,0.011537,0.114451,0.058918,652475.0
6,log_incwage,elig_12_18,0.004896,0.007603,0.519559,0.058913,652475.0
7,log_incwage,cumulative_eligibility,0.008646,0.006476,0.181882,0.058917,652475.0
8,log_inctot,elig_0_5,0.017117,0.023316,0.462873,0.056865,726631.0
9,log_inctot,elig_6_11,0.010853,0.011343,0.338664,0.056866,726631.0


### Summary of Findings (Labor Market Outcomes)

This regression analysis aimed to assess the impact of childhood Medicaid eligibility on various labor market outcomes, controlling for childhood economic conditions, a linear time trend, and fixed effects for state, age, sex, race, and Hispanic origin.

*   **Employment (`employed`):**
    *   Medicaid eligibility in early childhood (0-5 years) shows a positive, but statistically insignificant, association with the likelihood of being employed.
    *   For the 12-18 age window (`elig_12_18`) and cumulative eligibility (`cumulative_eligibility`), there is a statistically significant **positive association** with the likelihood of being employed (p < 0.05). Specifically, a one-unit increase in `elig_12_18` is associated with approximately a 0.65 percentage point increase in the probability of being employed, and for cumulative eligibility, it's about 0.54 percentage points.

*   **Log Wage Income (`log_incwage`):**
    *   Most eligibility measures show no statistically significant impact on log wage income. The `elig_0_5` coefficient is negative, while `elig_6_11` and `cumulative_eligibility` are positive, but none are statistically significant at conventional levels.

*   **Log Total Income (`log_inctot`):**
    *   Cumulative eligibility (`cumulative_eligibility`) shows a statistically significant **positive association** with log total income (p < 0.01). A one-unit increase in cumulative eligibility is associated with approximately a 1.43 percent increase in total income.
    *   Eligibility during the 12-18 age window also shows a marginally significant positive association (p < 0.1).

*   **Weeks Worked Midpoint (`weeks_worked_midpoint`):**
    *   None of the Medicaid eligibility measures show a statistically significant impact on the weeks worked midpoint at conventional levels (e.g., p < 0.05 or p < 0.1), although `cumulative_eligibility` has a positive coefficient with a p-value of 0.094.

**Overall Interpretation:**

Unlike the educational outcomes where significant effects were largely absent or negative for "some college," this analysis reveals some positive and statistically significant associations between Medicaid eligibility and labor market outcomes. Specifically, increased eligibility, especially in later childhood (12-18) and cumulatively, appears to be associated with a higher probability of being employed and, for cumulative eligibility, higher total income. These findings suggest potential long-term benefits of early childhood health interventions on adult labor market participation and earnings.

## Final Task

### Subtask:
Provide a brief text summary of the regression findings, highlighting any statistically significant effects of Medicaid eligibility on labor market outcomes.


## Summary:

### Q&A
The regression analysis found statistically significant positive effects of Medicaid eligibility on certain labor market outcomes. Specifically, Medicaid eligibility during the 12-18 age window and cumulative eligibility are associated with an increased likelihood of being employed. Additionally, cumulative eligibility is significantly associated with higher total income. However, no statistically significant impact was observed on wage income or weeks worked at conventional levels.

### Data Analysis Key Findings
*   **Employment:** Medicaid eligibility during the 12-18 age window (`elig_12_18`) shows a statistically significant positive association with the likelihood of being employed (p < 0.05), increasing the probability by approximately 0.65 percentage points. Cumulative eligibility also has a statistically significant positive association (p < 0.05), increasing the probability by about 0.54 percentage points.
*   **Total Income:** Cumulative Medicaid eligibility (`cumulative_eligibility`) is significantly associated with an increase in log total income (p < 0.01), corresponding to an approximate 1.43 percent increase in total income for each unit increase in cumulative eligibility.
*   **Wage Income:** None of the Medicaid eligibility measures showed a statistically significant impact on log wage income at conventional levels.
*   **Weeks Worked:** No statistically significant impact was found for any Medicaid eligibility measure on the weeks worked midpoint at conventional levels (e.g., p < 0.05 or p < 0.1).

### Insights or Next Steps
*   The positive and significant associations between Medicaid eligibility (especially later childhood and cumulative) and adult employment and total income suggest potential long-term economic benefits of early childhood health interventions, which contrasts with the findings for educational outcomes.
*   Future research could explore the causal pathways through which sustained or later childhood Medicaid eligibility translates into improved labor market outcomes, considering potential mediators such as health status, educational attainment, or other human capital investments.


# Task
**Run WLS regression analysis for educational outcomes**

Iterate through the four educational outcomes (`hs_grad`, `college_grad`, `years_of_education`, `some_college`) and the four eligibility measures (`cumulative_eligibility`, `elig_0_5`, `elig_6_11`, `elig_12_18`). For each combination, execute the `run_wls_regression` function using `df_acs_clean_merged` to collect the coefficient, standard error, p-value, number of observations (N), and R-squared. Consolidate all regression results into a pandas DataFrame, add a column for significance stars (*** if p < 0.01, ** if p < 0.05, * if p < 0.1), and display the full results table. Finally, provide a brief text summary of the regression findings for educational outcomes, highlighting any statistically significant coefficients.

## Run Education Regressions

### Subtask:
Iterate through the educational outcomes and eligibility measures, running WLS regression for each combination.


**Reasoning**:
I will initialize an empty list, define the outcome and eligibility variables, then loop through all combinations, run the `run_wls_regression` function, append the results to the list, convert it to a DataFrame, and display it.



In [18]:
import pandas as pd

# Initialize a list to store results
results_education = []

# Define educational outcomes and eligibility measures
outcomes = ['hs_grad', 'college_grad', 'years_of_education', 'some_college']
eligibility_measures = ['cumulative_eligibility', 'elig_0_5', 'elig_6_11', 'elig_12_18']

print("Starting educational outcome regression analysis...")

# Iterate through each outcome and eligibility measure
for outcome in outcomes:
    for eligibility in eligibility_measures:
        print(f"Running regression for Outcome: {outcome}, Eligibility: {eligibility}...")
        # Run the regression
        elig_coef, elig_se, elig_pvalue, rsquared, nobs = run_wls_regression(outcome, eligibility, df_acs_clean_merged)

        # Append the results
        results_education.append({
            'Outcome': outcome,
            'Eligibility Variable': eligibility,
            'Eligibility Coef': elig_coef,
            'Eligibility SE': elig_se,
            'Eligibility P-value': elig_pvalue,
            'R-squared': rsquared,
            'Nobs': nobs
        })

# Convert results to a DataFrame
results_education_df = pd.DataFrame(results_education)

# Display the results
print("\nEducational Outcome Regression Analysis Results:")
display(results_education_df)

Starting educational outcome regression analysis...
Running regression for Outcome: hs_grad, Eligibility: cumulative_eligibility...
Running regression for Outcome: hs_grad, Eligibility: elig_0_5...
Running regression for Outcome: hs_grad, Eligibility: elig_6_11...
Running regression for Outcome: hs_grad, Eligibility: elig_12_18...
Running regression for Outcome: college_grad, Eligibility: cumulative_eligibility...
Running regression for Outcome: college_grad, Eligibility: elig_0_5...
Running regression for Outcome: college_grad, Eligibility: elig_6_11...
Running regression for Outcome: college_grad, Eligibility: elig_12_18...
Running regression for Outcome: years_of_education, Eligibility: cumulative_eligibility...
Running regression for Outcome: years_of_education, Eligibility: elig_0_5...
Running regression for Outcome: years_of_education, Eligibility: elig_6_11...
Running regression for Outcome: years_of_education, Eligibility: elig_12_18...
Running regression for Outcome: some_coll

,Outcome,Eligibility Variable,Eligibility Coef,Eligibility SE,Eligibility P-value,R-squared,Nobs
0,hs_grad,cumulative_eligibility,-0.001717,0.001714,0.316590,0.024506,816735.0
1,hs_grad,elig_0_5,0.000558,0.007924,0.943860,0.024503,816735.0
2,hs_grad,elig_6_11,-0.001529,0.003006,0.611056,0.024504,816735.0
3,hs_grad,elig_12_18,-0.001583,0.001568,0.312878,0.024505,816735.0
4,college_grad,cumulative_eligibility,0.001683,0.002262,0.457032,0.062403,816735.0
5,college_grad,elig_0_5,0.002450,0.012419,0.843632,0.062403,816735.0
6,college_grad,elig_6_11,-0.000300,0.004774,0.949967,0.062403,816735.0
7,college_grad,elig_12_18,0.002072,0.002660,0.436201,0.062403,816735.0
8,years_of_education,cumulative_eligibility,-0.008757,0.012301,0.476507,0.069771,816497.0
9,years_of_education,elig_0_5,0.035321,0.062713,0.573290,0.069771,816497.0


**Reasoning**:
Now that the regressions have been run, I will add a 'Significance' column to the results DataFrame based on the p-value. Then, I will display the updated DataFrame and provide a summary of the findings as requested.



In [19]:
def add_significance_stars(p_value):
    if p_value < 0.01:
        return '***'
    elif p_value < 0.05:
        return '**'
    elif p_value < 0.1:
        return '*'
    else:
        return ''

results_education_df['Significance'] = results_education_df['Eligibility P-value'].apply(add_significance_stars)

print("Educational Outcome Regression Analysis Results (with Significance):")
display(results_education_df)


Educational Outcome Regression Analysis Results (with Significance):


,Outcome,Eligibility Variable,Eligibility Coef,Eligibility SE,Eligibility P-value,R-squared,Nobs,Significance
0,hs_grad,cumulative_eligibility,-0.001717,0.001714,0.316590,0.024506,816735.0,
1,hs_grad,elig_0_5,0.000558,0.007924,0.943860,0.024503,816735.0,
2,hs_grad,elig_6_11,-0.001529,0.003006,0.611056,0.024504,816735.0,
3,hs_grad,elig_12_18,-0.001583,0.001568,0.312878,0.024505,816735.0,
4,college_grad,cumulative_eligibility,0.001683,0.002262,0.457032,0.062403,816735.0,
5,college_grad,elig_0_5,0.002450,0.012419,0.843632,0.062403,816735.0,
6,college_grad,elig_6_11,-0.000300,0.004774,0.949967,0.062403,816735.0,
7,college_grad,elig_12_18,0.002072,0.002660,0.436201,0.062403,816735.0,
8,years_of_education,cumulative_eligibility,-0.008757,0.012301,0.476507,0.069771,816497.0,
9,years_of_education,elig_0_5,0.035321,0.062713,0.573290,0.069771,816497.0,
